pkg/kubelet/images/puller.go

```diff
@@ -34,42 +37,102 @@ type pullResult struct {
 }
 
 type imagePuller interface {
-	pullImage(context.Context, kubecontainer.ImageSpec, []v1.Secret, chan<- pullResult, *runtimeapi.PodSandboxConfig)
+	pullImage(context.Context, kubecontainer.ImageSpec, []v1.Secret, chan pullResult, *runtimeapi.PodSandboxConfig, v1.PullPolicy)
 }
 
 var _, _ imagePuller = &parallelImagePuller{}, &serialImagePuller{}
 
 type parallelImagePuller struct {
 	imageService kubecontainer.ImageService
 	tokens       chan struct{}
+	pulls        sync.Map
+}
+
+type pullBroker struct {
+	subscribers sync.Map
+}
+
+func (p *pullBroker) subscribe(ch chan pullResult) {
+	p.subscribers.Store(ch, struct{}{})
+}
+
+func (p *pullBroker) publish(res pullResult) {
+	p.subscribers.Range(func(k, v any) bool {
+		ch, ok := k.(chan pullResult)
+		if !ok {
+			return false
+		}
+
+		select {
+		case ch <- res:
+		default:
+		}
+
+		return true
+	})
 }
 
 func newParallelImagePuller(imageService kubecontainer.ImageService, maxParallelImagePulls *int32) imagePuller {
 	if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
-		return &parallelImagePuller{imageService, nil}
+		return &parallelImagePuller{imageService, nil, sync.Map{}}
 	}
-	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls)}
+	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}}
 }
 
-func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan<- pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig) {
+func pullKey(spec kubecontainer.ImageSpec) string {
+	res := []string{}
+	for _, annotation := range spec.Annotations {
+		res = append(res, annotation.Name+"-"+annotation.Value)
+	}
+	sort.Strings(res)
+	if spec.RuntimeHandler != "" {
+		res = append([]string{spec.RuntimeHandler}, res...)
+	}
+	if spec.Image != "" {
+		res = append([]string{spec.Image}, res...)
+	}
+	return strings.Join(res, "-")
+}
+
+func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig, pullPolicy v1.PullPolicy) {
 	go func() {
+		key := pullKey(spec)
+
+		// A pull policy of 'Always' will result in calling CRI's PullImage
+		// every time.
+		if pullPolicy != v1.PullAlways {
+			// Reuse existing pull broker if available
+			if v, ok := pip.pulls.Load(key); ok {
+				existingPullBroker, _ := v.(*pullBroker)
+				existingPullBroker.subscribe(pullChan)
+				return
+			}
+		}
+
 		if pip.tokens != nil {
 			pip.tokens <- struct{}{}
 			defer func() { <-pip.tokens }()
 		}
+
+		// Create a new pull broker.
+		br := &pullBroker{}
+		br.subscribe(pullChan)
+		pip.pulls.Store(key, br)
```

Comment:
can we leak memory if two images raced to this point and both do `Store` of their broker with the same key?

Reply:
Yes, but I can fix that by using `LoadOrStore`.